src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,305 @@
+        final double throttledSamplingRate = samplingRate
+                * (1.0 - greenbidsAnalyticsProperties.getExploratorySamplingSplit());
+
+        final long hashInt = Math.abs(greenbidsId.hashCode()) % 0x10000;
+
+        final boolean isPrimarySampled = hashInt < exploratorySamplingRate * 0xFFFF;
+        if (isPrimarySampled) {
+            return true;
+        }
+        return hashInt >= (1 - throttledSamplingRate) * 0xFFFF;
+    }
+
+    public CommonMessage createBidMessage(
+            AuctionContext auctionContext,
+            BidResponse bidResponse
+    ) {
+        final List<Imp> imps = auctionContext.getBidRequest().getImp();
+
+        if (imps == null || imps.isEmpty()) {
+            throw new IllegalArgumentException("imps is null or empty");
+        }
+
+        final Long auctionTimestamp = Optional.of(auctionContext.getBidRequest())
+                .map(BidRequest::getExt)
+                .map(ExtRequest::getPrebid)
+                .map(ExtRequestPrebid::getAuctiontimestamp)
+                .orElse(null);
+
+        final long auctionElapsed = auctionTimestamp != null ? System.currentTimeMillis() - auctionTimestamp : 0L;
+
+        final Map<String, Bid> seatsWithBids = Optional.ofNullable(bidResponse.getSeatbid())
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatBid -> !seatBid.getBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatBid::getSeat,
+                                seatBid -> seatBid.getBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final List<SeatNonBid> seatNonBids = auctionContext.getBidRejectionTrackers().entrySet().stream()
+                .map(entry -> toSeatNonBid(entry.getKey(), entry.getValue()))
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .toList();
+
+        final Map<String, NonBid> seatsWithNonBids = Optional.of(seatNonBids)
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatNonBid::getSeat,
+                                seatNonBid -> seatNonBid.getNonBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final List<AdUnit> adUnitsWithBidResponses = extractAdUnitsWithBidResponses(
+                imps, seatsWithBids, seatsWithNonBids
+        );
+
+        return CommonMessage.builder()
+                .version(greenbidsAnalyticsProperties.getAnalyticsServerVersion())
+                .auctionId(auctionContext.getBidRequest().getId())
+                .referrer(auctionContext.getBidRequest().getSite().getPage())
```

Comment:
general comment: make sure the chains of getters doesn't lead to the NPE (I mean, if you are sure it's safe, then it's fine, otherwise it's better to handle such cases, for example, with Optional or straightforward null check)

Reply:
Have added Optional chains on the following occurences getters
- auctionId
- referrer
- imps
- banner: width, height, pos, id